### ISSUE TITLE
Different date formats in SQLite database for 11/30/23 and 12/01/23 (following day)

### ISSUE THREAD

**Issue by thomasWuppermann** [murrayo/yaspe#3]
I importe html files for mutlitple days into one database which worked quite well. Only the time format for vmstat and mgstat metrics differ between the files for November and those for December. The html files from System Performance Tool contain the same format. Why does YASPE change the format?
Now it is hard to create SQL queries for metrics for both months. 

For clarification:
- html-files contain the dates for mgstat and vmstat in the format: MM/DD/YY HH:mm:ss
- SQLite-DB contains the dates for mgstat and vmstat **for November** as follows: **MM/DD/YY** HH:mm:ss
- SQLite-DB contains the dates for mgstat and vmstat **for December** as follows: **DD/MM/YY** HH:mm:ss

**Comment by murrayo** [comment]
Hi, this sounds like a bug. I recently fixed a previous bug around figuring out whether 1/12/23 is 1st December or 12 Jan. It seems that introduced another bug! The action of yaspe is intended to convert dates to all the same format. e.g. dd/mm/yy or mm/dd/yy. Because pButtons or SystemPerformance data from around the world comes in many formats this is a harder problem than it first looks like. I have some test data. And hope to look at this in the next few days. Do you use a container? Are you using Intel or a MAC M1? 
Thanks

**Comment by thomasWuppermann** [comment]
Hi, thanks for your answer. Now I understand the problem. I'm using docker on Windows / Intel for converting / extracting the data. Would it be a solution or at least a workaround to add an optional parameter with the format of the timestamp for those cases where it is difficult or nearly impossible to figure out the right format? Not so convenient but secure...
Thanks! Thomas

**Comment by murrayo** [comment]
yes. I have thought of that, and that may end up being the last-chance solution to override my dodgy programming logic ;) 
Depending on my workload I will look at this ASAP. I will make sure the windows/Intel image is updated.

**Comment by cliffmason** [comment]
Hi Murray –

Not sure how this got on my personal email, but there have been enough date-related issues (why can’t the world have a single standard) that you really may have to do a format field to replace what you have.

Cliff

**Comment by thomasWuppermann** [comment]
To clarify my use-case (which might be the same for others):
for uor customer we are using System Performance Tools for multiple Days / a few weeks to examine some problems in a multi-stage mirrored IRIS environment. To help us with the huge amount of data we use YASPE - SQLITE - Grafana to visualize some of the data which is great!
Besides the problem descibed above we now have another problem with different systems creating different formats for iostat. Some habe 12-hour-AM/PM-Format, some 24 hour format. Since SQLite does not now anything about date-formats and grafana needs the same date-format for every data I would have to insert another manual step to change the format in those SQLite-DB containing AM/PM-formatted dates.

For the SQLite-Export it would be best if 
1. all dates in the different tables would be in the same format
2. one could determine the format so it would be the same over multiple databases.

**Comment by murrayo** [comment]
Hi, sorry about the delay. I have made a fix to the Intel docker image. the source here on Git Hub is also updated. This needs some more work, however, could you test? I will be working on this some more in the new year. Thanks

**Comment by thomasWuppermann** [comment]
Great, thanks!
I will be testing this today or tomorrow and give my feedback.
Greatings, Thomas

Von: murrayo ***@***.***>
Gesendet: Donnerstag, 21. Dezember 2023 07:25
An: murrayo/yaspe ***@***.***>
Cc: Thomas Wuppermann | HBT ***@***.***>; Author ***@***.***>
Betreff: Re: [murrayo/yaspe] Different date formats in SQLite database for 11/30/23 and 12/01/23 (following day) (Issue #3)


Hi, sorry about the delay. I have made a fix to the Intel docker image. the source here on Git Hub is also updated. This needs some more work, however, could you test? I will be working on this some more in the new year. Thanks

**Comment by thomasWuppermann** [comment]
Hi,
works pretty well for me!
There are not different datetime formats in one file anymore. Different formats between systems and / or types (mgstat, iostat or vmstat) are still there, but I can handle that.
Since I now use csv-files rather then sqlite-db  and convert them to openmetrics-format to backfill then into Prometheus I have to convert (“re-order”) the timestamps anyways.
As for now I can work with this but if you want to enhance the tool, I will be happy to test it.
Best regards, Thomas

Von: murrayo ***@***.***>
Gesendet: Donnerstag, 21. Dezember 2023 07:25
An: murrayo/yaspe ***@***.***>
Cc: Thomas Wuppermann | HBT ***@***.***>; Author ***@***.***>
Betreff: Re: [murrayo/yaspe] Different date formats in SQLite database for 11/30/23 and 12/01/23 (following day) (Issue #3)


Hi, sorry about the delay. I have made a fix to the Intel docker image. the source here on Git Hub is also updated. This needs some more work, however, could you test? I will be working on this some more in the new year. Thanks

**Comment by murrayo** [comment]
Did you know there is a "-D" option to force date formats to DDMMYYYY for csv file output? I will extend this to the database.

**Comment by murrayo** [comment]
...I will make the date format selectable.

**Comment by thomasWuppermann** [comment]
Great! And no, didn’t know this. I just tried the following but with no luck. As for now I can work with what you changed very well so better prepare for the holydays 😉
What I did:
docker run -v $workdir:/data --rm --name yaspe murray1oldfield/yaspe:latest ./yaspe.py -i /data/<htmlFile> -a -s -x -o multi -c -D

A warning I got for every iostat.csv-file (but the files were fine, just the same date-format as without “-D”
/app/./yaspe.py:203: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.

Best regards,
Thomas

Von: murrayo ***@***.***>
Gesendet: Donnerstag, 21. Dezember 2023 20:33
An: murrayo/yaspe ***@***.***>
Cc: Thomas Wuppermann | HBT ***@***.***>; Author ***@***.***>
Betreff: Re: [murrayo/yaspe] Different date formats in SQLite database for 11/30/23 and 12/01/23 (following day) (Issue #3)


Did you know there is a "-D" option to force date formats to DDMMYYYY for csv file output? I will extend this to the database.

**Comment by thomasWuppermann** [comment]
Would be great but – as I said – think of the holydays coming 😉

Von: murrayo ***@***.***>
Gesendet: Donnerstag, 21. Dezember 2023 20:33
An: murrayo/yaspe ***@***.***>
Cc: Thomas Wuppermann | HBT ***@***.***>; Author ***@***.***>
Betreff: Re: [murrayo/yaspe] Different date formats in SQLite database for 11/30/23 and 12/01/23 (following day) (Issue #3)


...I will make the date format selectable.

**Comment by thomasWuppermann** [comment]
Hi,

just to let you know:
I began to use YASPE again and to dig a bit deeper….
The -D option is fine and helps with the output but if YASPE cannot determine the format of the input (for example on January first) the output can still be wrong.
But I don’t see a chance to determine the date format fully automatically especially for January 1st to 12th and some other dates.

I tried a few things with data from December 31st to January 2nd
Time-format in the html-files is DD/MM/YYYY but in the csv-file it changes from MM/DD/YYYY für December to MM/DD/YYYY for January 2nd.
(I still don’t know if the format is consistent during one run of YASPE because I only have html-files for 24 hours and on January 1st you can’t see which is month and which is day ).

As for now I’m able to reach my goal “converting the csv-files to openmetrics format” by
distinguishing between 2023 and 2024 for the creation of unix timestamp.

That’s it for now.

Best regards, and thanks again, YASPE is great 😉

Thomas



Von: murrayo ***@***.***>
Gesendet: Donnerstag, 21. Dezember 2023 20:33
An: murrayo/yaspe ***@***.***>
Cc: Thomas Wuppermann | HBT ***@***.***>; Author ***@***.***>
Betreff: Re: [murrayo/yaspe] Different date formats in SQLite database for 11/30/23 and 12/01/23 (following day) (Issue #3)


Did you know there is a "-D" option to force date formats to DDMMYYYY for csv file output? I will extend this to the database.

**Comment by murrayo** [comment]
Hi, I refactored the date handling code and tested using some sample files I have that roll over months or year-end. When you have time please test with the files you had a problem with and let me know :) 
The containers have not been updated. I will run for a few days before I do that.

**Comment by murrayo** [comment]
Also FYI the date stored in the SQLite file is now always YYYY/mm/dd

**Comment by cliffmason** [comment]
Murray,

I think this was Ken with the issue.  And maybe Ray (I can’t remember).  Works fine on my stuff.  Updated yaspe anyway.

Cliff

**Comment by thomasWuppermann** [comment]
Hi, that is great, I’ve seen this but had no time to test it yet, sorry.
As soon as there is an hour left, I will test it and tell my conclusions.
Great tool and thanks for your effort!



Von: murrayo ***@***.***>
Gesendet: Dienstag, 13. Februar 2024 07:04
An: murrayo/yaspe ***@***.***>
Cc: Thomas Wuppermann | HBT ***@***.***>; Author ***@***.***>
Betreff: Re: [murrayo/yaspe] Different date formats in SQLite database for 11/30/23 and 12/01/23 (following day) (Issue #3)


Also FYI the date stored in the SQLite file is now always YYYY/mm/dd
